src/core/segment_reader.rs

```diff
@@ -280,6 +282,94 @@ impl SegmentReader {
         Ok(inv_idx_reader)
     }
 
+    /// Returns the list of fields that have been indexed in the segment.
+    /// The field list includes the field defined in the schema as well as the fields
+    /// that have been indexed as a part of a JSON field.
+    /// The returned field name is the full field name, including the name of the JSON field.
+    pub fn fields_metadata(&self) -> crate::Result<Vec<FieldMetadata>> {
+        let mut indexed_fields: Vec<(String, Type)> = Vec::new();
+        for (field, field_entry) in self.schema().fields() {
+            let field_name = field_entry.name().to_string();
+            let is_indexed = field_entry.is_indexed();
+
+            if is_indexed {
+                indexed_fields.push((
+                    field_name.to_string(),
+                    field_entry.field_type().value_type(),
+                ));
+                let is_json = field_entry.field_type().value_type() == Type::Json;
+                if is_json {
+                    let inv_index = self.inverted_index(field)?;
+                    let encoded_fields_in_index = inv_index.list_encoded_fields()?;
+                    indexed_fields.extend(
+                        encoded_fields_in_index
+                            .into_iter()
+                            .map(|(name, typ)| (format!("{}.{}", field_name, name), typ)),
```

Comment:
I suspect having picked `(String, Type)` here, forcing the concatenation of the field and the field path was not a good idea. This is my first pass reviewing so I might edit this comment later.

Reply:
Second pass here.
I think keep the `field_name` and the `field_path` separated might be helpful. The trouble is that the fast fields are all in the same bag so I am not sure this is a good idea.
So your Vec would become `Vec<(&str, Option<&str>, Type)>`.
The stored field logic would become computationally simpler.
The fast field logic would then get the extra split logic.

Also quite frankly, I am not entirely sure this is better... You can ignore this comment.

Reply:
We need to return the full name, as this is used e.g. to select a field by a user